### PR TITLE
Feat.cast.loc

### PR DESCRIPTION
**Pull request by aneksteind**
Separating some improvements for #788, namely
- keying casts by `Location`
- expecting casts from void to be the first statement in the destination block of an allocation call
- ensuring that the argument to `free` is not modified between its cast and its use

### CODE REVIEW

File: c2rust-analyze/src/c_void_casts.rs [hunk 1]
```diff
@@ -347,14 +362,28 @@ impl<'tcx> CVoidCasts<'tcx> {
                     To => args[0].place().unwrap(),
                 };
                 let c_void_ptr = CVoidPtr::checked(c_void_ptr, &body.local_decls, tcx);
-                let get_cast =
-                    move |stmt: &Statement<'tcx>| c_void_ptr.get_cast_from_stmt(direction, stmt);
+                let get_cast = move |(idx, stmt): (usize, &Statement<'tcx>)| {
```
Comment (aneksteind) [inline c2rust-analyze/src/c_void_casts.rs:365]: @kkysen I don't really like this, and it was a function of needing type annotations for this closure in `find_last_cast_curr_block` for example. I'm open to cleaner ways of going about this.

File: c2rust-analyze/src/borrowck/type_check.rs [hunk 2]
```diff
@@ -440,17 +436,21 @@ pub fn visit_body<'tcx>(
...
-            tc.current_location = Location {
-                block: bb,
+            let loc = Location {
+                block,
                 statement_index: idx,
             };
-            tc.visit_statement(stmt);
+            tc.current_location = loc;
+
+            if !c_void_casts.should_skip_stmt(&loc) {
```
Comment (aneksteind) [inline c2rust-analyze/src/borrowck/type_check.rs:447]: I did this to avoid passing `loc` into `visit_statement`, but if it's inappropriate to check `c_void_casts` here I can move it into `visit_statement`

File: c2rust-analyze/src/c_void_casts.rs [hunk 3]
```diff
@@ -158,25 +162,33 @@ pub struct CVoidCast<'tcx> {
...
+            .0
+            .get(loc)
+            .map(|(_void, subst)| subst)
+            .unwrap_or(&place)
     }
 
-    pub fn insert(&mut self, cast: CVoidCast<'tcx>) {
-        self.0.insert(cast.c_void_ptr, cast.other_ptr);
+    pub fn insert(&mut self, loc: Location, cast: CVoidCast<'tcx>) {
+        self.0.insert(loc, (cast.c_void_ptr, cast.other_ptr));
```
Comment (spernsteiner) [inline c2rust-analyze/src/c_void_casts.rs:191]: This is maybe excessively cautious, but could you assert here than `loc` isn't already in the map?  This should always be the case since we currently have at most one "from" cast and at most one "to" cast per `Location`.

File: c2rust-analyze/src/c_void_casts.rs [hunk 4]
```diff
@@ -158,25 +162,33 @@ pub struct CVoidCast<'tcx> {
...
-        *self.0.get(&place).unwrap_or(&place)
+    pub fn get_adjusted_place_or_default_to(
+        &self,
+        loc: &Location,
+        place: Place<'tcx>,
+    ) -> Place<'tcx> {
+        *self
+            .0
+            .get(loc)
+            .map(|(_void, subst)| subst)
```
Comment (spernsteiner) [inline c2rust-analyze/src/c_void_casts.rs:186]: Similarly, you could assert here than `void == place`

File: c2rust-analyze/src/c_void_casts.rs [hunk 5]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
-            None => return false,
-            Some(it) => it,
-        };
-        self.to.contains(lhs)
-            || self.from.contains(match get_cast_place(rv) {
-                None => return false,
-                Some(it) => it,
-            })
+    pub fn should_skip_stmt(&self, loc: &Location) -> bool {
+        self.to.contains(loc) || self.from.contains(loc)
```
Comment (spernsteiner) [inline c2rust-analyze/src/c_void_casts.rs:291]: Can you please add a separate field to `CVoidCastsUniDirectional` for tracking "casts to skip", separate from the map for tracking "calls with a cast on the argument/result"?  Right now, an entry in the `CVoidCastsUniDirectional` map means "either this is a call, and has the following cast on its argument/result, OR this is a cast, and it should be skipped".  This will work, but IMO it would be clearer to have separate maps/sets for the two cases.

File: c2rust-analyze/src/c_void_casts.rs [hunk 6]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+    }
+
+    fn is_place_local_modified(p: &Place, stmt: &Statement) -> bool {
+        if !p.projection.is_empty() {
+            return false;
+        }
```
Comment (spernsteiner) [inline c2rust-analyze/src/c_void_casts.rs:297]: What is this function checking, exactly?  I'm not sure whether this first case should return `false` because `p` isn't a local, or `true` because we can't be sure that `stmt` leaves `p` unmodified.

File: c2rust-analyze/src/c_void_casts.rs [hunk 7]
```diff
@@ -320,41 +405,44 @@ impl<'tcx> CVoidCasts<'tcx> {
...
-                let c_void_ptr = match direction {
+                let c_void_ptr_place = match direction {
                     From => destination,
                     To => args[0].place().unwrap(),
                 };
-                let c_void_ptr = CVoidPtr::checked(c_void_ptr, &body.local_decls, tcx);
+                let c_void_ptr = CVoidPtr::checked(c_void_ptr_place, &body.local_decls, tcx);
                 let get_cast =
-                    move |stmt: &Statement<'tcx>| c_void_ptr.get_cast_from_stmt(direction, stmt);
+                    |stmt: &Statement<'tcx>| c_void_ptr.get_cast_from_stmt(direction, stmt);
```
Comment (spernsteiner) [inline c2rust-analyze/src/c_void_casts.rs:419]: Since this is the only callback ever passed to `find_*_cast_*_block`, would it make sense to pass `c_void_ptr` and `stmt` to those functions instead, and inline this `get_cast_from_stmt` call?

**Review by spernsteiner — APPROVED** [review]

File: c2rust-analyze/src/c_void_casts.rs (hunk 6 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs:297]: In particular it was meant to check what your [comment](https://github.com/immunant/c2rust/pull/788#discussion_r1113481984) outlined:

> I think we can conservatively treat the Place as unmodified only if (1) it's just a Local, with no projections, and (2) that local isn't mentioned in the LHS or RHS of any statements between the cast and the call.

File: c2rust-analyze/src/c_void_casts.rs (hunk 7 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs:419]: I don't mind either way, I'll play around with it to see if one appears to be better than the other.

File: c2rust-analyze/src/c_void_casts.rs [hunk 8]
```diff
@@ -158,25 +162,33 @@ pub struct CVoidCast<'tcx> {
 /// to their properly typed pointers,
 /// like [`CVoidCasts`], but in a single direction, meaning it represents either
 /// [`CVoidCastDirection::From`] or [`CVoidCastDirection::To`].
-#[derive(Default)]
-pub struct CVoidCastsUniDirectional<'tcx>(HashMap<CVoidPtr<'tcx>, Place<'tcx>>);
+#[derive(Default, Clone, Debug)]
+pub struct CVoidCastsUniDirectional<'tcx>(HashMap<Location, (CVoidPtr<'tcx>, Place<'tcx>)>);
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:166]: `(CVoidPtr<'tcx>, Place<'tcx>)` can just be `CVoidCast<'tcx>`.  They were only split before due to being the key and value of the `HashMap`.

File: c2rust-analyze/src/c_void_casts.rs [hunk 9]
```diff
@@ -158,25 +162,33 @@ pub struct CVoidCast<'tcx> {
...
 
 impl<'tcx> CVoidCastsUniDirectional<'tcx> {
-    pub fn contains(&self, place: Place<'tcx>) -> bool {
-        self.0.contains_key(&place)
+    pub fn contains(&self, loc: &Location) -> bool {
+        self.0.contains_key(loc)
     }
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:171]: For here and other places, you can just take `Location` by value as it's `Copy`.

File: c2rust-analyze/src/c_void_casts.rs (hunk 6 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs:297]: Can you add that as docs?

File: c2rust-analyze/src/c_void_casts.rs [hunk 10]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+        false
+    }
+
+    fn find_first_cast_succ_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        // For [`CVoidCastDirection::From`], we only count
+        // a cast from `*c_void` to an arbitrary type in the subsequent block,
+        // searching forward.
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:340]: Make this rustdoc.  It was only a local comment before since it was in a closure.

File: c2rust-analyze/src/c_void_casts.rs [hunk 11]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+    }
+
+    fn find_last_cast_curr_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+        place: &Place<'tcx>,
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        // For [`CVoidCastDirection::From`], we only count
+        // a cast to `*c_void` from an arbitrary type in the same block,
+        // searching backwards.
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:362]: Same for these docs.

File: c2rust-analyze/src/c_void_casts.rs [hunk 12]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+        }
+
+        None
+    }
+
+    fn find_last_cast_curr_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+        place: &Place<'tcx>,
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:359]:
```suggestion
    fn find_last_cast_curr_block<F>(
        mut get_cast: F,
        statements: &[Statement<'tcx>],
        place: &Place<'tcx>,
    ) -> Option<(usize, CVoidCast<'tcx>)>
    where
        F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>,
    {
```

File: c2rust-analyze/src/c_void_casts.rs [hunk 13]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+                Cast(_, op, _) => op.place(),
+                // BinaryOp
+                // CheckedBinaryOp
+                // NullaryOp
+                UnaryOp(_, op) => op.place(),
+                Discriminant(p) => Some(p),
+                // Aggregate
+                ShallowInitBox(op, _) => op.place(),
+                _ => None,
+            };
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:324]: If you're naming all the variants by comments anyways, just include them as actual arms and no `_ =>`.

File: c2rust-analyze/src/c_void_casts.rs [hunk 14]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+
+    fn is_place_local_modified(p: &Place, stmt: &Statement) -> bool {
+        if !p.projection.is_empty() {
+            return false;
+        }
+
+        let local = p.local;
+
+        if let StatementKind::Assign(asgn) = stmt.kind.clone() {
+            let (lhs, rv) = *asgn;
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:302]: nit:
```suggestion
        if let StatementKind::Assign(assign) = stmt.kind.clone() {
            let (lhs, rv) = *assign;
```

File: c2rust-analyze/src/c_void_casts.rs (hunk 13 above)
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:324]:
```suggestion
        if let StatementKind::Assign(assign) = &stmt.kind {
            let (lhs, rv) = &**assign;
            if lhs.local == local {
                return true;
            }

            use Rvalue::*;
            let rv_place = match rv {
                Use(op) => op.place(),
                Repeat(op, _) => op.place(),
                Ref(_, _, p) => Some(*p),
                // ThreadLocalRef
                AddressOf(_, p) => Some(*p),
                Len(p) => Some(*p),
                Cast(_, op, _) => op.place(),
                // BinaryOp
                // CheckedBinaryOp
                // NullaryOp
                UnaryOp(_, op) => op.place(),
                Discriminant(p) => Some(*p),
                // Aggregate
                ShallowInitBox(op, _) => op.place(),
                _ => None,
            };
```

Don't need to clone `StatementKind` and its `Box`.

File: c2rust-analyze/src/c_void_casts.rs (hunk 10 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs:340]: The function is different now, though, so can you update the docs accordingly, too?

File: c2rust-analyze/src/c_void_casts.rs (hunk 7 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs:419]: I only added the closure initially to avoid passing `direction` and making the code a bit shorter.  You can switch to whatever's simpler now.

File: c2rust-analyze/src/c_void_casts.rs [hunk 15]
```diff
@@ -303,7 +388,7 @@ impl<'tcx> CVoidCasts<'tcx> {
     ///
     /// [`*c_void`]: core::ffi::c_void
     fn insert_all_from_body(&mut self, body: &Body<'tcx>, tcx: TyCtxt<'tcx>) {
-        for bb_data in body.basic_blocks().iter() {
+        for (block, bb_data) in body.basic_blocks().iter().enumerate() {
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:391]:
```suggestion
        for (block, bb_data) in body.basic_blocks().iter_enumerated() {
```

Then you don't need the `BasicBlock::from_usize(block)`s.

File: c2rust-analyze/src/c_void_casts.rs [hunk 16]
```diff
@@ -245,16 +257,17 @@ impl<'tcx> CVoidCasts<'tcx> {
...
     ) -> Place<'tcx> {
         self.direction(direction)
-            .get_adjusted_place_or_default_to(place)
+            .get_adjusted_place_or_default_to(loc, place)
     }
 
     /// See [`CVoidCastsUniDirectional::insert`].
-    fn insert(&mut self, direction: CVoidCastDirection, cast: CVoidCast<'tcx>) {
-        self.direction_mut(direction).insert(cast)
+    fn insert(&mut self, loc: Location, direction: CVoidCastDirection, cast: CVoidCast<'tcx>) {
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:269]: I'd switch `loc` and `direction`.  That's the first lookup before `location`.

File: c2rust-analyze/src/c_void_casts.rs [hunk 17]
```diff
@@ -245,16 +257,17 @@ impl<'tcx> CVoidCasts<'tcx> {
     /// See [`CVoidCastsUniDirectional::get_adjusted_place_or_default_to`].
     pub fn get_adjusted_place_or_default_to(
         &self,
+        loc: &Location,
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:260]: Same here about switching `loc` and `direction`.

File: c2rust-analyze/src/c_void_casts.rs [hunk 18]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+    }
+
+    fn find_first_cast_succ_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        // For [`CVoidCastDirection::From`], we only count
+        // a cast from `*c_void` to an arbitrary type in the subsequent block,
+        // searching forward.
+        for (sidx, stmt) in statements.iter().enumerate() {
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:341]: You can use `.iter_enumerated()` to keep the index typed.

File: c2rust-analyze/src/c_void_casts.rs [hunk 19]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+
+    fn find_last_cast_curr_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+        place: &Place<'tcx>,
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        // For [`CVoidCastDirection::From`], we only count
+        // a cast to `*c_void` from an arbitrary type in the same block,
+        // searching backwards.
+        for (sidx, stmt) in statements.iter().enumerate().rev() {
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:363]: `.iter_enumerated()` again.

File: c2rust-analyze/src/c_void_casts.rs (hunk 18 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs:341]: Nevermind, I didn't realize these were just normal slices.

File: c2rust-analyze/src/c_void_casts.rs [hunk 20]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+                continue;
+            }
+            if let Some(cast) = get_cast(stmt) {
+                return Some((sidx, cast));
+            } else {
+                break;
+            }
+        }
+
+        None
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:352]:
```suggestion
        statements
            .iter()
            .enumerate()
            .skip_while(|(_, stmt)| matches!(stmt.kind, StatementKind::StorageDead(_)))
            .find_map(|(i, stmt)| Some((i, get_cast(stmt)?)))
```

I think the intent is a lot clearer in a functional style.  `.skip_while` and `.find_map` work very well here.

Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs:352]: Written this way, docs are much less needed, though I still think they'd be useful.

File: c2rust-analyze/src/c_void_casts.rs [hunk 21]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+    }
+
+    fn find_last_cast_curr_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+        place: &Place<'tcx>,
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:358]: `Place` is copy, so you can take it by value.

**Review by kkysen — COMMENTED** [review]
Can you make sure to update the docs when things are changed, and include new ones for non-trivial functions?  Also, can you put in the PR description what's being fixed?  Probably just Stuart's comment from #788.

File: c2rust-analyze/src/c_void_casts.rs [hunk 22]
```diff
@@ -320,41 +405,44 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+                            From => target.unwrap(),
+                        },
+                    };
+                    self.insert(loc, direction, cast.clone());
+
+                    let loc = Location {
+                        statement_index: bb_data.statements.len(),
+                        block: BasicBlock::from_usize(block),
+                    };
+                    self.insert(loc, direction, cast);
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:445]: Can you inline `loc` both times?  It's a bit confusing to reuse the variable name when it's totally different.

Also, why do we need to insert the cast at two `Location`s?  Can you document why that's done?

File: c2rust-analyze/src/c_void_casts.rs [hunk 23]
```diff
@@ -320,41 +405,44 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+                        statement_index,
+                        block: match direction {
+                            To => BasicBlock::from_usize(block),
+                            From => target.unwrap(),
+                        },
+                    };
+                    self.insert(loc, direction, cast.clone());
+
+                    let loc = Location {
+                        statement_index: bb_data.statements.len(),
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:442]: What does it mean to have an out-of-bounds `statement_index` for a `Location`?

File: c2rust-analyze/src/c_void_casts.rs [hunk 24]
```diff
@@ -320,41 +405,44 @@ impl<'tcx> CVoidCasts<'tcx> {
...
                 .copied()
             {
                 use CVoidCastDirection::*;
-                let c_void_ptr = match direction {
+                let c_void_ptr_place = match direction {
                     From => destination,
                     To => args[0].place().unwrap(),
                 };
-                let c_void_ptr = CVoidPtr::checked(c_void_ptr, &body.local_decls, tcx);
+                let c_void_ptr = CVoidPtr::checked(c_void_ptr_place, &body.local_decls, tcx);
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:417]: This was supposed to shadow the first `c_void_ptr`, now `c_void_ptr_place`.  Why the change?

File: c2rust-analyze/src/c_void_casts.rs [hunk 25]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
-        self.to.contains(lhs)
-            || self.from.contains(match get_cast_place(rv) {
-                None => return false,
-                Some(it) => it,
-            })
+    pub fn should_skip_stmt(&self, loc: &Location) -> bool {
+        self.to.contains(loc) || self.from.contains(loc)
+    }
+
+    fn is_place_local_modified(p: &Place, stmt: &Statement) -> bool {
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:294]:
```suggestion
    fn is_place_unmodified_by_stmt(p: &Place, stmt: &Statement) -> bool {
```

I think this was meant to be "unmodified", right?  Since Stuart said it's
> unmodified only if (1) it's just a `Local`, with no projections, and
and here it's returning `false` when there is a projection.

Also, I think `is_place_unmodified_by_stmt` reads more clearly.

File: c2rust-analyze/src/c_void_casts.rs (hunk 5 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs:291]: Is this why there are two insertions into `c_void_casts`?  I agree it would be much clearer to separate the two.

File: c2rust-analyze/src/c_void_casts.rs [hunk 26]
```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
...
+        // searching backwards.
+        for (sidx, stmt) in statements.iter().enumerate().rev() {
+            let cast = get_cast(stmt);
+            if let Some(cast) = cast {
+                return Some((sidx, cast));
+            } else if Self::is_place_local_modified(place, stmt) {
+                return None;
+            }
+        }
+        None
```
Comment (kkysen) [inline c2rust-analyze/src/c_void_casts.rs:371]: Can you explain (in docs, too) what this is doing?

File: c2rust-analyze/src/c_void_casts.rs (hunk 23 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs:442]: I think a function like this in `util.rs` would make things a lot clearer.  I'd've thought it'd be in `rustc` already.

```rust
pub fn terminator_location(block: BasicBlock, block_data: &BasicBlockData) -> Location {
    Location {
        block,
        statement_index: block_data.statements.len(),
    }
}
```

File: c2rust-analyze/src/c_void_casts.rs (hunk 3 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs:192]: 91bad4f2

File: c2rust-analyze/src/c_void_casts.rs (hunk 4 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: c27a916f

File: c2rust-analyze/src/c_void_casts.rs (hunk 7 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: b6e492d4

File: c2rust-analyze/src/c_void_casts.rs (hunk 4 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs]: nit: you can use `assert_eq!`.

File: c2rust-analyze/src/c_void_casts.rs (hunk 5 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: d2cd89bf

File: c2rust-analyze/src/c_void_casts.rs (hunk 10 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 8fcc2a8a

File: c2rust-analyze/src/c_void_casts.rs (hunk 9 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 7fb2d448

File: c2rust-analyze/src/c_void_casts.rs (hunk 11 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 8fcc2a8a

File: c2rust-analyze/src/c_void_casts.rs (hunk 8 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: bc23fd71

File: c2rust-analyze/src/c_void_casts.rs (hunk 13 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: a8c64a5b

File: c2rust-analyze/src/c_void_casts.rs (hunk 10 above)
Comment (kkysen) [reply c2rust-analyze/src/c_void_casts.rs]: These two functions are a bit different now, like this one with skipping the `StorageDead`s at the beginning, can you explain that in the docs?  The other one is a bit more complex, so that one is more in need of updated docs.

File: c2rust-analyze/src/c_void_casts.rs (hunk 13 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 4bb545f4

File: c2rust-analyze/src/c_void_casts.rs (hunk 20 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: f62cb76a

File: c2rust-analyze/src/c_void_casts.rs (hunk 22 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: inlined in 54830958

File: c2rust-analyze/src/c_void_casts.rs (hunk 24 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: f6741d57

File: c2rust-analyze/src/c_void_casts.rs (hunk 6 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 93a30394

File: c2rust-analyze/src/c_void_casts.rs (hunk 25 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 93a30394, it's checking for modification, not unmodification

File: c2rust-analyze/src/c_void_casts.rs (hunk 26 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: ed3c93a9

File: c2rust-analyze/src/c_void_casts.rs (hunk 23 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: a7ddc515

File: c2rust-analyze/src/c_void_casts.rs (hunk 16 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 65d1bce3

File: c2rust-analyze/src/c_void_casts.rs (hunk 15 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 89fc7fd9

File: c2rust-analyze/src/c_void_casts.rs (hunk 4 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 59dffcc3

File: c2rust-analyze/src/c_void_casts.rs (hunk 10 above)
Comment (aneksteind) [reply c2rust-analyze/src/c_void_casts.rs]: 827606fc
